app/app.go

```diff
@@ -71,9 +72,35 @@ func (a *App) handleSpans(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusAccepted)
 }
 
+// ungzipWrap wraps a handleFunc and transparently ungzips the body of the
+// request if it is gzipped
+func ungzipWrap(hf func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var newBody io.ReadCloser
+		isGzipped := r.Header.Get("Content-Encoding")
+		if isGzipped == "gzip" {
+			buf := bytes.Buffer{}
+			if _, err := io.Copy(&buf, r.Body); err != nil {
+				w.WriteHeader(http.StatusBadRequest)
```

Comment:
Might want to log something in addition to returning an error to the client, here and below.